### PR TITLE
Adding URLExtractor class

### PR DESCRIPTION
**Pull request by m-cheung**
This pull request attempts to resolve issue #40

### CODE REVIEW

File: lib/org/chromium/webidl/URLExtractor.js
Comment (m-cheung) [inline lib/org/chromium/webidl/URLExtractor.js]: I've modified the RegExp slightly from the original implementation. This one performs the following:
`http` followed by optional `s`, followed by `://`, `www.` is optional.
Labels can be up to 256 characters (each individually is 63, but overall is 256) and currently matching `[-_a-zA-Z0-9.]`. Followed by mandatory `.` and suffix of 2-6 alpha characters.

- Paths and queries are matches in a similar manner as before

Comment (arobins) [reply lib/org/chromium/webidl/URLExtractor.js]: How restrictive does this need to be? The code that uses the output of this already needs to deal with invalid stuff (because your pattern matches 'http://.......com/').

File: test/any/URLExtractor-test.js
Comment (arobins) [inline test/any/URLExtractor-test.js]: I think this will fail if you had anything after it on the same line.

File: test/any/URLExtractor-test.js
Comment (arobins) [inline test/any/URLExtractor-test.js]: Put two URLs on the same line somewhere (in this test, or as a separate test). The extractor should work if the input is a single line.

File: test/any/URLExtractor-test.js
Comment (arobins) [inline test/any/URLExtractor-test.js]: The pattern has support for anchors. That should be tested. It might be better as a separate test because of how this one works though.

File: test/any/URLExtractor-test.js
Comment (arobins) [inline test/any/URLExtractor-test.js]: You don't use the last value in the array.

File: test/any/URLExtractor-test.js
Comment (arobins) [inline test/any/URLExtractor-test.js]: Seeing this makes me think that the extractor should support an array of patterns.

File: lib/org/chromium/webidl/URLExtractor.js
Comment (m-cheung) [reply lib/org/chromium/webidl/URLExtractor.js]: I believe it just needs to be sufficient to extract well-formed urls from IDL files. The previous regex was `https?:\/\/[^/]+(\/[^?#, \n]*)?(\?[^#, \n]*)?` but this was even less restrictive and resulted in matches of weird results (thus the change in this revision)

I can attempt at writing a stricter regex (maybe `https?:\/\/(www\.)?([a-zA-Z0-9][-_a-zA-Z0-9])([-_a-zA-Z0-9.])*\.[a-z]{2,6}(\/[^?#, \n]*)?(\?[^#, \n]*)?` would be a better option?)

File: test/any/URLExtractor-test.js
Comment (m-cheung) [reply test/any/URLExtractor-test.js]: The pattern should only match up to the anchor tag. Everything after the anchor tag should be dropped (the anchor doesn't matter since the entire page is fetched anyways). I will add a test in the next commit to verify this.

File: test/any/URLExtractor-test.js
Comment (m-cheung) [reply test/any/URLExtractor-test.js]: Do you think an array of Javascript RegExp patterns is better or an array of RegExp strings?
(e.g.  `[ /potato\.com/, /.*\.google\.com/]` vs `[ 'potato\.com', '.*\.google\.com' ]` )

File: test/any/URLExtractor-test.js
Comment (m-cheung) [reply test/any/URLExtractor-test.js]: Thanks for catching that. It has been addressed.

File: lib/org/chromium/webidl/URLExtractor.js
Comment (arobins) [reply lib/org/chromium/webidl/URLExtractor.js]: Do you have examples of 'weird' results that the old pattern was catching that it shouldn't?

We have a URL parser somewhere, because HTTP fetches wouldn't work without it. Why wouldn't we use it to do validation after using a simple regex for the first pass instead of trying to write a regex for URL syntax?

File: test/any/URLExtractor-test.js
Comment (arobins) [reply test/any/URLExtractor-test.js]: I can hardly even read JS, but I think array of patterns makes more sense.

File: lib/org/chromium/webidl/URLExtractor.js
Comment (m-cheung) [reply lib/org/chromium/webidl/URLExtractor.js]: I will have to double check, but if I recall correctly, the original pattern fails a couple of tests because it matches stuff like
```
http://google.com
www.google.com
```
as one URL. I can rerun the test with the original pattern later and post the observations here.

With regards to the URL parser, Node.js has URL parsers built in and it is being used by FOAM when attempting to perform the fetch. However, we currently do not do validation on it. We can perform an additional URL parse and validation if you feel that this may be a good idea.

Comment (m-cheung) [reply lib/org/chromium/webidl/URLExtractor.js]: The test named 'should return all HTTP(s) URLs within the text fragment' will returning the following results when executed with the original Regexp
```
[ "http://www.google.com Some other thing\n        potato https://www.google.com/",
"//www.google.com/",
null ]
```

Comment (arobins) [reply lib/org/chromium/webidl/URLExtractor.js]: The original pattern doesn't know how to deal with a domain-only URL. This probably works though:
```
https?:\/\/[^/?#, \n]+(\/[^?#, \n]*)?(\?[^#, \n]*)?
```
Though now that I've seen it, I think I like this even better:
```
https?:\/\/[^#, \n]
```
because delegating the decision of what is and isn't a valid URL to some other code feels like the right thing to do.

Comment (m-cheung) [reply lib/org/chromium/webidl/URLExtractor.js]: I have taken your suggestion and changed the regex to `https?:\/\/[^#, \n]*`. We will delegate the decision of what is and isn't a valid URL to FOAM HTTPRequest

**Comment by m-cheung** [comment]
All requests should be resolved by now and should be ready for another round of review.

File: lib/org/chromium/webidl/URLExtractor.js
Comment (arobins) [inline lib/org/chromium/webidl/URLExtractor.js]: This should probably be private

**Review by arobins — COMMENTED** [review]
lgtm, but @mdittmer should approve

File: lib/org/chromium/webidl/URLExtractor.js
Comment (m-cheung) [reply lib/org/chromium/webidl/URLExtractor.js]: I have changed it to `excludeRegexp_` and `includeRegexp_` respectively. I believe that's the FOAM way of representing internal properties, but I could be wrong. @mdittmer please chime in if it is something else

File: lib/org/chromium/webidl/URLExtractor.js
Comment (mdittmer) [inline lib/org/chromium/webidl/URLExtractor.js]: No need to comment this out. Also, what are `expressions` here? Regular expressions?

File: lib/org/chromium/webidl/URLExtractor.js
Comment (mdittmer) [inline lib/org/chromium/webidl/URLExtractor.js]: Don't comment out.

File: lib/org/chromium/webidl/URLExtractor.js
Comment (mdittmer) [inline lib/org/chromium/webidl/URLExtractor.js:1]: Let's switch the `postSet`s in this file to expressions:

```js
{ name: 'foo' },
{ name: 'barDepsOnFoo_', expression: function(foo) { /* ... */ return barDepsOnFooValue; } }
```

The name of the parameter on the expression's `function` must match the property name(s) it depends on.

**Review by mdittmer — APPROVED** [review]
LGTM after issues addressed. Poke me offline if you have questions about replacing `postSet`s with `expression`s.

**Pull request merged**
